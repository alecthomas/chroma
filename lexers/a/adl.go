package a

import (
	. "github.com/alecthomas/chroma" // nolint
	"github.com/alecthomas/chroma/lexers/internal"
)

// ADL lexer.
var Adl = internal.Register(MustNewLexer(
	&Config{
		Name:      "ADL",
		Aliases:   []string{"adl"},
		Filenames: []string{"*.adl", "*.adls", "*.adlf", "*.adlx"},
	},
	Rules{
		"root": {},
	},
))
