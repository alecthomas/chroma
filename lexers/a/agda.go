package a

import (
	. "github.com/alecthomas/chroma" // nolint
	"github.com/alecthomas/chroma/lexers/internal"
)

// Agda lexer.
var Agda = internal.Register(MustNewLexer(
	&Config{
		Name:      "Agda",
		Aliases:   []string{"agda"},
		Filenames: []string{"*.agda"},
		MimeTypes: []string{"text/x-agda"},
	},
	Rules{
		"root": {},
	},
))
