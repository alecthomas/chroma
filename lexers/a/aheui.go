package a

import (
	. "github.com/alecthomas/chroma" // nolint
	"github.com/alecthomas/chroma/lexers/internal"
)

// Aheui lexer.
var Aheui = internal.Register(MustNewLexer(
	&Config{
		Name:      "Aheui",
		Aliases:   []string{"aheui"},
		Filenames: []string{"*.aheui"},
	},
	Rules{
		"root": {},
	},
))
