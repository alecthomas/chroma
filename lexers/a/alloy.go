package a

import (
	. "github.com/alecthomas/chroma" // nolint
	"github.com/alecthomas/chroma/lexers/internal"
)

// Alloy lexer.
var Alloy = internal.Register(MustNewLexer(
	&Config{
		Name:      "Alloy",
		Aliases:   []string{"alloy"},
		Filenames: []string{"*.als"},
		MimeTypes: []string{"text/x-alloy"},
	},
	Rules{
		"root": {},
	},
))
