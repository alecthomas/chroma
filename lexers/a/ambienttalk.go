package a

import (
	. "github.com/alecthomas/chroma" // nolint
	"github.com/alecthomas/chroma/lexers/internal"
)

// AmbientTalk lexer.
var AmbientTalk = internal.Register(MustNewLexer(
	&Config{
		Name:      "AmbientTalk",
		Aliases:   []string{"at", "ambienttalk", "ambienttalk/2"},
		Filenames: []string{"*.at"},
		MimeTypes: []string{"text/x-ambienttalk"},
	},
	Rules{
		"root": {},
	},
))
