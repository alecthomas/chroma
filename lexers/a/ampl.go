package a

import (
	. "github.com/alecthomas/chroma" // nolint
	"github.com/alecthomas/chroma/lexers/internal"
)

// Ampl lexer.
var Ampl = internal.Register(MustNewLexer(
	&Config{
		Name:      "Ampl",
		Aliases:   []string{"ampl"},
		Filenames: []string{"*.run"},
	},
	Rules{
		"root": {},
	},
))
