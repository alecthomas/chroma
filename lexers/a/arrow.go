package a

import (
	. "github.com/alecthomas/chroma" // nolint
	"github.com/alecthomas/chroma/lexers/internal"
)

// Arrow lexer.
var Arrow = internal.Register(MustNewLexer(
	&Config{
		Name:      "Arrow",
		Aliases:   []string{"arrow"},
		Filenames: []string{"*.arw"},
	},
	Rules{
		"root": {},
	},
))
