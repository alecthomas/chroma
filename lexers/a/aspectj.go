package a

import (
	. "github.com/alecthomas/chroma" // nolint
	"github.com/alecthomas/chroma/lexers/internal"
)

// AspectJ lexer.
var AspectJ = internal.Register(MustNewLexer(
	&Config{
		Name:      "AspectJ",
		Aliases:   []string{"aspectj"},
		Filenames: []string{"*.aj"},
		MimeTypes: []string{"text/x-aspectj"},
	},
	Rules{
		"root": {},
	},
))
