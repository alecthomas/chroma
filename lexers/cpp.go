package lexers

import (
	"regexp"

	. "github.com/alecthomas/chroma/v2" // nolint
)

var (
	cppAnalyserIncludeRe   = regexp.MustCompile(`#include <[a-z_]+>`)
	cppAnalyserNamespaceRe = regexp.MustCompile(`using namespace `)
)

var CPP = Register(MustNewXMLLexer(
	embedded,
	"embedded/c++.xml",
).SetConfig(
	&Config{
		Name:      "C++",
		Aliases:   []string{"cpp", "c++"},
		Filenames: []string{"*.cpp", "*.hpp", "*.c++", "*.h++", "*.cc", "*.hh", "*.cxx", "*.hxx", "*.C", "*.H", "*.cp", "*.CPP", "*.cppm", "*.ixx", "*.tpp"},
		MimeTypes: []string{"text/x-c++hdr", "text/x-c++src"},
		Priority:  0.1,
		EnsureNL:  true,
	},
)).SetAnalyser(func(text string) float32 {
	if cppAnalyserIncludeRe.MatchString(text) {
		return 0.2
	}

	if cppAnalyserNamespaceRe.MatchString(text) {
		return 0.4
	}

	return 0
})
