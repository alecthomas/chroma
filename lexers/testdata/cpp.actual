#include "a"
#include <b>

[[nodiscard]] void foo() noexcept;
void foo();

constexpr class {} a;

enum class E { A, B };
enum class [[nodiscard]] E { A, B };

enum class {
  a, b,
  c,
} e;

enum E { A, B };

class A {
  void foo();
  void bar();
};

int main() {
    return 0 + 1'3 + 1.4;
}
