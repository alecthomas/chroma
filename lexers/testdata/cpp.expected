[
  {"type":"CommentPreproc","value":"#include"},
  {"type":"Text","value":" "},
  {"type":"CommentPreprocFile","value":"\"a\""},
  {"type":"CommentPreproc","value":"\n#include"},
  {"type":"Text","value":" "},
  {"type":"CommentPreprocFile","value":"\u003cb\u003e"},
  {"type":"CommentPreproc","value":"\n"},
  {"type":"Text","value":"\n"},
  {"type":"NameAttribute","value":"[[nodiscard]]"},
  {"type":"Text","value":" "},
  {"type":"KeywordType","value":"void"},
  {"type":"Text","value":" "},
  {"type":"Name","value":"foo"},
  {"type":"Punctuation","value":"()"},
  {"type":"Text","value":" "},
  {"type":"Keyword","value":"noexcept"},
  {"type":"Punctuation","value":";"},
  {"type":"Text","value":"\n"},
  {"type":"KeywordType","value":"void"},
  {"type":"Text","value":" "},
  {"type":"NameFunction","value":"foo"},
  {"type":"Punctuation","value":"();"},
  {"type":"Text","value":"\n\n"},
  {"type":"Keyword","value":"constexpr"},
  {"type":"Text","value":" "},
  {"type":"Keyword","value":"class"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"{}"},
  {"type":"Text","value":" "},
  {"type":"Name","value":"a"},
  {"type":"Punctuation","value":";"},
  {"type":"Text","value":"\n\n"},
  {"type":"Keyword","value":"enum"},
  {"type":"Text","value":" "},
  {"type":"Keyword","value":"class"},
  {"type":"Text","value":" "},
  {"type":"NameClass","value":"E"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"Text","value":" "},
  {"type":"Name","value":"A"},
  {"type":"Punctuation","value":","},
  {"type":"Text","value":" "},
  {"type":"Name","value":"B"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"};"},
  {"type":"Text","value":"\n"},
  {"type":"Keyword","value":"enum"},
  {"type":"Text","value":" "},
  {"type":"Keyword","value":"class"},
  {"type":"Text","value":" "},
  {"type":"NameAttribute","value":"[[nodiscard]]"},
  {"type":"Text","value":" "},
  {"type":"NameClass","value":"E"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"Text","value":" "},
  {"type":"Name","value":"A"},
  {"type":"Punctuation","value":","},
  {"type":"Text","value":" "},
  {"type":"Name","value":"B"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"};"},
  {"type":"Text","value":"\n\n"},
  {"type":"Keyword","value":"enum"},
  {"type":"Text","value":" "},
  {"type":"Keyword","value":"class"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"Text","value":"\n  "},
  {"type":"Name","value":"a"},
  {"type":"Punctuation","value":","},
  {"type":"Text","value":" "},
  {"type":"Name","value":"b"},
  {"type":"Punctuation","value":","},
  {"type":"Text","value":"\n  "},
  {"type":"Name","value":"c"},
  {"type":"Punctuation","value":","},
  {"type":"Text","value":"\n"},
  {"type":"Punctuation","value":"}"},
  {"type":"Text","value":" "},
  {"type":"Name","value":"e"},
  {"type":"Punctuation","value":";"},
  {"type":"Text","value":"\n\n"},
  {"type":"Keyword","value":"enum"},
  {"type":"Text","value":" "},
  {"type":"NameClass","value":"E"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"Text","value":" "},
  {"type":"Name","value":"A"},
  {"type":"Punctuation","value":","},
  {"type":"Text","value":" "},
  {"type":"Name","value":"B"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"};"},
  {"type":"Text","value":"\n\n"},
  {"type":"Keyword","value":"class"},
  {"type":"Text","value":" "},
  {"type":"NameClass","value":"A"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"Text","value":"\n  "},
  {"type":"KeywordType","value":"void"},
  {"type":"Text","value":" "},
  {"type":"NameFunction","value":"foo"},
  {"type":"Punctuation","value":"();"},
  {"type":"Text","value":"\n  "},
  {"type":"KeywordType","value":"void"},
  {"type":"Text","value":" "},
  {"type":"NameFunction","value":"bar"},
  {"type":"Punctuation","value":"();"},
  {"type":"Text","value":"\n"},
  {"type":"Punctuation","value":"};"},
  {"type":"Text","value":"\n\n"},
  {"type":"KeywordType","value":"int"},
  {"type":"Text","value":" "},
  {"type":"NameFunction","value":"main"},
  {"type":"Punctuation","value":"()"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"Text","value":"\n    "},
  {"type":"Keyword","value":"return"},
  {"type":"Text","value":" "},
  {"type":"LiteralNumberInteger","value":"0"},
  {"type":"Text","value":" "},
  {"type":"Operator","value":"+"},
  {"type":"Text","value":" "},
  {"type":"LiteralNumberInteger","value":"1'3"},
  {"type":"Text","value":" "},
  {"type":"Operator","value":"+"},
  {"type":"Text","value":" "},
  {"type":"LiteralNumberFloat","value":"1.4"},
  {"type":"Punctuation","value":";"},
  {"type":"Text","value":"\n"},
  {"type":"Punctuation","value":"}"},
  {"type":"Text","value":"\n"}
]
